hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/interfaces/BlockManager.java

```diff
@@ -55,12 +55,25 @@ long putBlock(Container container, BlockData data, boolean endOfBlock)
    *
    * @param container - Container from which block needs to be fetched.
    * @param blockID - BlockID of the Block.
+   * @param isReplicaCheckRequired - if true will match container data's replica index with provided BlockID.
    * @return Block Data.
    * @throws IOException when BcsId is unknown or mismatched
    */
-  BlockData getBlock(Container container, BlockID blockID)
+  BlockData getBlock(Container container, BlockID blockID, boolean isReplicaCheckRequired)
       throws IOException;
```

Comment:
I thought since the function verifys BcsID, it would be logical to verifyReplicaIndex as well. Since like how BCSId is for Ratis. ReplicaIndex is for EC.

Reply:
I see your point and having those two verifications in the same place has some merit.  But bcsID is always verified, while replicaIndex verification is conditional on yet another parameter.  So it is entirely the responsibility of the caller.  I think this, and the need to increase the size of the `BlockManager` interface makes having the check inside `getBlock` worse than having it outside.

Reply:
Unfortunately in today's date this is conditional because of supporting the older clients. We can eventually remove this check when we deprecate older clients and always perform the replica index check

Reply:
This would atleast make the caller aware that there is a replicaIndex check that would happen and would not lead to future regressions.

Reply:
@adoroszlai What do you think about this? I have made the other changes, just want to know what your opinion about this is. I believe we should have this check inside getBlock function to avoid future regressions. But yeah I am fine with making the change if you feel otherwise.

Reply:
> Unfortunately in today's date this is conditional because of the need to support the older clients. We can eventually remove this check when we deprecate older clients and always perform the replica index check

If we keep the check out of `getBlock` now, we can move it into `getBlock` when the time comes to always perform the check.

The overloaded method with extra `boolean` parameter would need to be deprecated if the check is added in `getBlock` now.

> I believe we should have this check inside getBlock function to avoid future regressions

The method is overloaded, and the default value for the new parameter (passed from the old method) is `false`.  So it is just as easy to introduce regression by calling the old `getBlock` as it would be by forgetting to call `verifyReplicaIndex` before calling `getBlock`.

Reply:
Should we just make the default value true and then pass false in other places where it is being used now?

Reply:
I have made the change and moved the check outside and removed the boolean parameter

Reply:
If we would like to make `verifyReplicaIdx` and `verifyBCSId` calls consistent (i.e. happen in the same places), we should move `verifyBCSId` out of `getBlock`.

Current call hierarchy:

```
verifyBCSId
> BlockManagerImpl.getBlock
  > BlockManagerImpl.getBlock
    > FilePerChunkStrategy.readChunk
      > KeyValueHandler.handleGetSmallFile
      > KeyValueHandler.handleReadChunk
    > KeyValueHandler.handleGetSmallFile
  > KeyValueHandler.handleGetBlock
> KeyValueHandler.handleGetCommittedBlockLength
> KeyValueHandler.handleReadChunk
```

So all callers originate from 4 methods of `KeyValueHandler`.  I think that's where verification should happen.  Some of them already do, as shown above.

Verifying either `bcsID` or `replicaIndex` multiple times for the same `handleReadChunk` operation is unnecessary.

`handleGetSmallFile` calls the old `getBlock`, but I think it should also `verifyReplicaIndex`.

So the two `verify...` calls only need to be added in `handleGetBlock` and `handleGetSmallFile`.

Reply:
Can handleGetSmall have EC replication?

Reply:
Since there is no EC replication there we are going to be fetching the data as it is we don't need the verify replica index for Small files

Reply:
We can take up this refactoring task item as part of another patch.